pkg/scm/git/git.go

```diff
@@ -217,84 +220,98 @@ func makePathAbsolute(source string) string {
 // expect git clone spec syntax; it also provides details if the file://
 // proto was explicitly specified, if we should use OS copy vs. the git
 // binary, and if a frag/ref has a bad format
-func ParseFile(source string) (details *FileProtoDetails, mods *URLMods) {
+func ParseFile(source string) (*FileProtoDetails, *URLMods, error) {
+	// Checking to see if the user included a "file://" in the call
 	protoSpecified := false
 	if strings.HasPrefix(source, "file://") && len(source) > 7 {
 		protoSpecified = true
 	}
 
+	refSpecified := false
+	path, ref := "", ""
+	if strings.LastIndex(source, "#") != -1 {
+		refSpecified = true
+
+		segments := strings.SplitN(source, "#", 2)
+		path = segments[0]
+		ref = segments[1]
+	} else {
+		path = source
+	}
+
 	// in each valid case, like the prior logic in scm.go did, we'll make the
 	// paths absolute and prepend file:// to the path which callers should
 	// switch to
+	if doesExist(path) {
```

Comment:
this is what i was trying to guide you towards.   I think it retains all the logic we've discussed, but is more compact:

```
	// in each valid case, like the prior logic in scm.go did, we'll make the
	// paths absolute and prepend file:// to the path which callers should
	// switch to
	if doesExist(path) {

		// Is there even a valid .git repository?
		isValidGit, err := isValidGitRepository(path)
		hasGit:=false
		if isValidGit {
		  hasGit=hasGitBinary()
		}
		// err indicates empty git repo (should also consider throwing an error for no .git folder
		// so the caller knows specifically why it's not a valid git repo)
		if err != nil || !isValidGit || !hasGit {
			details := &FileProtoDetails{
				UseCopy:        true,
				FileExists:     true,
				BadRef:         false,  
				ProtoSpecified: protoSpecified,
			}
			mods := &URLMods{
				Scheme: "file",
				Path:   makePathAbsolute(strings.TrimPrefix(path, "file://")),
				Ref:    ref,
			}
			return details, mods, err
		}

		// Check is the #ref is valid
		badRef:=refSpecified && !gitSSHURLPathRef.MatchString(ref)

		details := &FileProtoDetails{
			BadRef:         badRef,
			FileExists:     true,
			ProtoSpecified: protoSpecified,
			// this value doesn't really matter, we should not proceed if the git ref is bad
			// but let's fallback to "copy" mode if the ref is invalid.
			UseCopy:        badRef,  
		}

		mods := &URLMods{
			Scheme: "file",
			Path:   makePathAbsolute(strings.TrimPrefix(path, "file://")),
			Ref:    ref,
		}
		return details, mods, nil
	}

	// File does not exist, return bad
	details := &FileProtoDetails{
		UseCopy:        false,
		FileExists:     false,
		BadRef:         false,
		ProtoSpecified: protoSpecified,
	}
	return details, nil, nil
```